GPU/Debugger/Record.cpp

```diff
@@ -0,0 +1,1013 @@
+		lastRegisters.clear();
+
+		commands.push_back(last);
+	}
+}
+
+static std::string GenRecordingFilename() {
+	const std::string dumpDir = GetSysDirectory(DIRECTORY_DUMP);
+	const std::string prefix = dumpDir + "/" + g_paramSFO.GetDiscID();
+
+	File::CreateFullPath(dumpDir);
+
+	for (int n = 1; n < 10000; ++n) {
+		std::string filename = StringFromFormat("%s_%04d.ppdmp", prefix.c_str(), n);
+		if (!File::Exists(filename)) {
+			return filename;
+		}
+	}
+
+	return StringFromFormat("%s_%04d.ppdmp", prefix.c_str(), 9999);
+}
+
+static void EmitDisplayBuf() {
+	struct DisplayBufData {
+		PSPPointer<u8> topaddr;
+		u32 linesize, pixelFormat;
+	};
+
+	DisplayBufData disp{};
+	__DisplayGetFramebuf(&disp.topaddr, &disp.linesize, &disp.pixelFormat, 0);
+
+	u32 ptr = (u32)pushbuf.size();
+	u32 sz = (u32)sizeof(disp);
+	pushbuf.resize(pushbuf.size() + sz);
+	memcpy(pushbuf.data() + ptr, &disp, sz);
+
+	commands.push_back({CommandType::DISPLAY, sz, ptr});
+}
+
+static void BeginRecording() {
+	u32 ptr = (u32)pushbuf.size();
+	u32 sz = 512 * 4;
+	pushbuf.resize(pushbuf.size() + sz);
+	gstate.Save((u32_le *)(pushbuf.data() + ptr));
+
+	commands.push_back({CommandType::INIT, sz, ptr});
+}
+
+static void WriteCompressed(FILE *fp, const void *p, size_t sz) {
+	size_t compressed_size = snappy_max_compressed_length(sz);
+	u8 *compressed = new u8[compressed_size];
+	snappy_compress((const char *)p, sz, (char *)compressed, &compressed_size);
+
+	u32 write_size = (u32)compressed_size;
+	fwrite(&write_size, sizeof(write_size), 1, fp);
+	fwrite(compressed, compressed_size, 1, fp);
+
+	delete [] compressed;
+}
+
+static void WriteRecording() {
+	FlushRegisters();
+	EmitDisplayBuf();
+
+	const std::string filename = GenRecordingFilename();
+	FILE *fp = File::OpenCFile(filename, "wb");
+	fwrite(HEADER, sizeof(HEADER), 1, fp);
+	fwrite(&VERSION, sizeof(VERSION), 1, fp);
+
+	u32 sz = (u32)commands.size();
+	fwrite(&sz, sizeof(sz), 1, fp);
+	u32 bufsz = (u32)pushbuf.size();
+	fwrite(&bufsz, sizeof(bufsz), 1, fp);
+
+	WriteCompressed(fp, commands.data(), commands.size() * sizeof(Command));
+	WriteCompressed(fp, pushbuf.data(), bufsz);
+
+	fclose(fp);
+}
+
+static void GetVertDataSizes(int vcount, const void *indices, u32 &vbytes, u32 &ibytes) {
+	VertexDecoder vdec;
+	VertexDecoderOptions opts{};
+	vdec.SetVertexType(gstate.vertType, opts);
+
+	if (indices) {
+		u16 lower = 0;
+		u16 upper = 0;
+		GetIndexBounds(indices, vcount, gstate.vertType, &lower, &upper);
+
+		vbytes = (upper + 1) * vdec.VertexSize();
+		u32 idx = gstate.vertType & GE_VTYPE_IDX_MASK;
+		if (idx == GE_VTYPE_IDX_8BIT) {
+			ibytes = vcount * sizeof(u8);
+		} else if (idx == GE_VTYPE_IDX_16BIT) {
+			ibytes = vcount * sizeof(u16);
+		} else if (idx == GE_VTYPE_IDX_32BIT) {
+			ibytes = vcount * sizeof(u32);
+		}
+	} else {
+		vbytes = vcount * vdec.VertexSize();
+	}
+}
+
+static const u8 *mymemmem(const u8 *haystack, size_t hlen, const u8 *needle, size_t nlen) {
```

Comment:
Yeah, not cross platform.  At least doesn't seem to exist in MSVC.

-[Unknown]